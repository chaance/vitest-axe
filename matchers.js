export * from "./dist/matchers.js";
