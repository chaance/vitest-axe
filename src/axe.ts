import type AxeCore from "axe-core";
import { merge } from "lodash-es";
import { isHTMLElement, isHTMLString } from "./utils";
import { createRequire } from "node:module";

const require = createRequire(import.meta.url);
// eslint-disable-next-line @typescript-eslint/consistent-type-imports
const axeCore: typeof import("axe-core") = require("axe-core");
const { configure, run } = axeCore;

const AXE_RULES_COLOR = axeCore.getRules(["cat.color"]);

/**
 * Converts a HTML string or HTML element to a mounted HTML element.
 * @param html HTML element or a HTML string
 * @returns HTML element and a function to restore the document
 */
function mount(html: Element | string): [HTMLElement, () => void] {
	if (isHTMLElement(html)) {
		if (document.body.contains(html)) {
			return [html, () => undefined];
		}
		html = html.outerHTML;
	}

	if (isHTMLString(html)) {
		let originalHTML = document.body.innerHTML;
		function restore() {
			document.body.innerHTML = originalHTML;
		}

		document.body.innerHTML = html;
		return [document.body, restore];
	}

	if (typeof html === "string") {
		throw new Error(`html parameter ("${html}") has no elements`);
	}

	throw new Error(`html parameter should be an HTML string or an HTML element`);
}

/**
 * Small wrapper for axe-core#run that enables promises, default options and
 * injects html to be tested
 *
 * @param options default options to use in all instances
 *   - `globalOptions`: Global axe-core configuration (See
 *     https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#api-name-axeconfigure)
 *   - Any other property will be passed as the runner configuration (See
 *     https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#options-parameter)
 * @returns Instance of axe
 */
function configureAxe(
	options: AxeCore.RunOptions & {
		globalOptions?: AxeCore.Spec;
		impactLevels?: Array<AxeCore.ImpactValue>;
	} = {},
): (
	html: Element | string,
	additionalOptions?: AxeCore.RunOptions,
) => Promise<AxeCore.AxeResults> {
	let { globalOptions = {}, ...runnerOptions } = options;

	// Set the global configuration for axe-core
	// https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#api-name-axeconfigure
	const { rules = [], ...otherGlobalOptions } = globalOptions;

	// Color contrast checking doesnt work in a jsdom environment. So we need to
	// identify them and disable them by default.
	const defaultRules = AXE_RULES_COLOR.map(({ ruleId: id }) => ({
		id,
		enabled: false,
	}));

	configure({
		rules: [...defaultRules, ...rules],
		...otherGlobalOptions,
	});

	/**
	 * Small wrapper for axe-core#run that enables promises, default options and
	 * injects html to be tested
	 *
	 * @param html requires a html string to be injected into the body
	 * @param additionalOptions aXe options to merge with default options
	 * @returns Promise that will resolve with axe-core#run results object
	 */
	return function axe(
		html: Element | string,
		additionalOptions: AxeCore.RunOptions = {},
	): Promise<AxeCore.AxeResults> {
		let [element, restore] = mount(html);
		let options: AxeCore.RunOptions = merge(
			{},
			runnerOptions,
			additionalOptions,
		);

		return new Promise<AxeCore.AxeResults>((resolve) => {
			run(element, options, (err, results) => {
				restore();
				if (err) throw err;
				resolve(results);
			});
		});
	};
}

const axe = configureAxe();
export { configureAxe, axe };
export type { AxeCore };
