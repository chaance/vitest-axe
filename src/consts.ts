const HORIZONTAL_LINE = "\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500"
const LINE_BREAK = "\n\n"
const MIDDLE_DOT = "\u{00B7}"

export { HORIZONTAL_LINE, LINE_BREAK, MIDDLE_DOT }
