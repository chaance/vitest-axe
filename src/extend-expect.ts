/* eslint-disable @typescript-eslint/no-unused-vars */
import type { AxeMatchers } from "./matchers";

declare global {
  namespace Vi {
    interface Assertion<T = any> extends AxeMatchers {}
    interface AsymmetricMatchersContaining extends AxeMatchers {}
  }
}
