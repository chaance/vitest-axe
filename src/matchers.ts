import { toHaveNoViolations } from "./to-have-no-violations"
import type { AxeMatchers, NoViolationsMatcherResult } from "./types"

export { toHaveNoViolations }
export type { NoViolationsMatcherResult, AxeMatchers }
