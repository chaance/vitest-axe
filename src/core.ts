import axe from "axe-core"

export { axe as AxeCore }
