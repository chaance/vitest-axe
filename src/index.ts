import type { AxeMatchers } from "./matchers"
import { configureAxe, axe } from "./axe"
import type { AxeCore } from "./axe"

export { axe, configureAxe }
export type { AxeCore, AxeMatchers }
