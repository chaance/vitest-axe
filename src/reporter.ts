import chalk from "chalk"
import type { AxeCore } from "./core"
import { LINE_BREAK, HORIZONTAL_LINE } from "./consts"
import { printReceived } from "./utils"

const violationResolveHelpText = (violation: Pick<AxeCore.Result, "helpUrl">) => {
  return `You can find more information on this issue here: \n${chalk.blue(violation.helpUrl)}`
}

const expectedText = (node: AxeCore.NodeResult) => {
  const selector = node.target.join(", ")
  return `Expected the HTML found at $('${selector}') to have no violations:${LINE_BREAK}`
}

const reporter = (violations: AxeCore.Result[]) => {
  if (violations.length === 0) {
    return []
  }

  const delimiter = LINE_BREAK + HORIZONTAL_LINE + LINE_BREAK

  return violations
    .map((violation) => {
      const errorBody = violation.nodes
        .map((node) => {
          return (
            expectedText(node) +
            chalk.grey(node.html) +
            LINE_BREAK +
            `Received:` +
            LINE_BREAK +
            printReceived(`${violation.help} (${violation.id})`) +
            LINE_BREAK +
            chalk.yellow(node.failureSummary) +
            LINE_BREAK +
            (violation.helpUrl ? violationResolveHelpText(violation) : "")
          )
        })
        .join(LINE_BREAK)
      return errorBody
    })
    .join(delimiter)
}

export { reporter }
