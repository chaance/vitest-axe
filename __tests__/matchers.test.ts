import { describe, expect, it } from "vitest";
import { toHaveNoViolations } from "../src/matchers";
import { axe } from "../src/index";
import type { AxeCore } from "../src/index";

const PASSING_AXE_RESULTS: AxeCore.AxeResults = {
  testEngine: {
    name: "",
    version: "",
  },
  testRunner: {
    name: "",
  },
  testEnvironment: {
    userAgent: "",
    windowWidth: 0,
    windowHeight: 0,
  },
  url: "",
  timestamp: "",
  toolOptions: {},
  passes: [],
  violations: [],
  incomplete: [],
  inapplicable: [],
};
const FAILING_AXE_RESULTS: AxeCore.AxeResults = {
  ...PASSING_AXE_RESULTS,
  violations: [
    {
      id: "image-alt",
      impact: "critical",
      tags: [
        "cat.text-alternatives",
        "wcag2a",
        "wcag111",
        "section508",
        "section508.22.a",
      ],
      description:
        "Ensures <img> elements have alternate text or a role of none or presentation",
      help: "Images must have alternate text",
      helpUrl:
        "https://dequeuniversity.com/rules/axe/2.6/image-alt?application=axeAPI",
      nodes: [
        {
          any: [
            {
              id: "has-alt",
              data: null,
              relatedNodes: [],
              impact: "critical",
              message: "Element does not have an alt attribute",
            },
            {
              id: "aria-label",
              data: null,
              relatedNodes: [],
              impact: "serious",
              message: "aria-label attribute does not exist or is empty",
            },
            {
              id: "aria-labelledby",
              data: null,
              relatedNodes: [],
              impact: "serious",
              message:
                "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty or not visible",
            },
            {
              id: "non-empty-title",
              data: null,
              relatedNodes: [],
              impact: "serious",
              message:
                "Element has no title attribute or the title attribute is empty",
            },
            {
              id: "role-presentation",
              data: null,
              relatedNodes: [],
              impact: "minor",
              message:
                'Element\'s default semantics were not overridden with role="presentation"',
            },
            {
              id: "role-none",
              data: null,
              relatedNodes: [],
              impact: "minor",
              message:
                'Element\'s default semantics were not overridden with role="none"',
            },
          ],
          all: [],
          none: [],
          impact: "critical",
          html: '<img src="">',
          target: ["body > img"],
          failureSummary:
            'Fix any of the following:\n  Element does not have an alt attribute\n  aria-label attribute does not exist or is empty\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty or not visible\n  Element has no title attribute or the title attribute is empty\n  Element\'s default semantics were not overridden with role="presentation"\n  Element\'s default semantics were not overridden with role="none"',
        },
      ],
    },
  ],
};

describe("toHaveNoViolations", () => {
  it("returns a matcher function", () => {
    expect(toHaveNoViolations).toBeDefined();
    expect(typeof toHaveNoViolations).toBe("function");
  });

  it("throws error if non axe results object is passed", () => {
    expect(() => {
      // @ts-expect-error
      toHaveNoViolations({});
    }).toThrow("No violations found in aXe results object");
  });

  it("returns pass as true when no violations are present", () => {
    const matcherOutput = toHaveNoViolations(PASSING_AXE_RESULTS);
    expect(matcherOutput.pass).toBe(true);
  });

  it("returns same violations that are passed in the results object", () => {
    const matcherOutput = toHaveNoViolations(FAILING_AXE_RESULTS);
    expect(matcherOutput.actual).toBe(FAILING_AXE_RESULTS.violations);
  });

  it("returns correctly formatted message when violations are present", () => {
    const matcherOutput = toHaveNoViolations(FAILING_AXE_RESULTS);
    expect(typeof matcherOutput.message).toBe("function");
    expect(matcherOutput.message()).toMatchSnapshot();
  });

  it("returns pass as false when violations are present", () => {
    const matcherOutput = toHaveNoViolations(FAILING_AXE_RESULTS);
    expect(matcherOutput.pass).toBe(false);
  });

  it("returns properly formatted text with more complex example", async () => {
    const complexHtmlExample = `
        <html>
          <body>
            <a href="#link-name"></a>
            <a href="#link-name-2"></a>
            <img src="http://example.com"/>
            <img src="http://example.com/2"/>
            <input type="text"/>
          </body>
        </html>
      `;
    const results = await axe(complexHtmlExample);
    const matcherOutput = toHaveNoViolations(results);
    expect(matcherOutput.message()).toMatchSnapshot();
  });
});
