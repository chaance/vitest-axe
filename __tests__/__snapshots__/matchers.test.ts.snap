// Vitest Snapshot v1

exports[`toHaveNoViolations > returns correctly formatted message when violations are present 1`] = `
"<dim>expect(</><red>received</><dim>).toHaveNoViolations(</><green>expected</><dim>)</>

Expected the HTML found at $('body > img') to have no violations:

<gray><img src=\\"\\"></>

Received:

<red>\\"Images must have alternate text (image-alt)\\"</>

<yellow>Fix any of the following:</>
<yellow>  Element does not have an alt attribute</>
<yellow>  aria-label attribute does not exist or is empty</>
<yellow>  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty or not visible</>
<yellow>  Element has no title attribute or the title attribute is empty</>
<yellow>  Element's default semantics were not overridden with role=\\"presentation\\"</>
<yellow>  Element's default semantics were not overridden with role=\\"none\\"</>

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/2.6/image-alt?application=axeAPI</>"
`;

exports[`toHaveNoViolations > returns properly formatted text with more complex example 1`] = `
"<dim>expect(</><red>received</><dim>).toHaveNoViolations(</><green>expected</><dim>)</>

Expected the HTML found at $('img[src$=\\"example.com\\"]') to have no violations:

<gray><img src=\\"http://example.com\\"></>

Received:

<red>\\"Images must have alternate text (image-alt)\\"</>

<yellow>Fix any of the following:</>
<yellow>  Element does not have an alt attribute</>
<yellow>  aria-label attribute does not exist or is empty</>
<yellow>  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty</>
<yellow>  Element has no title attribute</>
<yellow>  Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"</>

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/4.4/image-alt?application=axeAPI</>

Expected the HTML found at $('img[src=\\"http://example.com/2\\"]') to have no violations:

<gray><img src=\\"http://example.com/2\\"></>

Received:

<red>\\"Images must have alternate text (image-alt)\\"</>

<yellow>Fix any of the following:</>
<yellow>  Element does not have an alt attribute</>
<yellow>  aria-label attribute does not exist or is empty</>
<yellow>  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty</>
<yellow>  Element has no title attribute</>
<yellow>  Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"</>

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/4.4/image-alt?application=axeAPI</>

────────

Expected the HTML found at $('input') to have no violations:

<gray><input type=\\"text\\"></>

Received:

<red>\\"Form elements must have labels (label)\\"</>

<yellow>Fix any of the following:</>
<yellow>  Form element does not have an implicit (wrapped) <label></>
<yellow>  Form element does not have an explicit <label></>
<yellow>  aria-label attribute does not exist or is empty</>
<yellow>  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty</>
<yellow>  Element has no title attribute</>
<yellow>  Element has no placeholder attribute</>
<yellow>  Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"</>

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/4.4/label?application=axeAPI</>

────────

Expected the HTML found at $('a[href$=\\"#link-name\\"]') to have no violations:

<gray><a href=\\"#link-name\\"></a></>

Received:

<red>\\"Links must have discernible text (link-name)\\"</>

<yellow>Fix all of the following:</>
<yellow>  Element is in tab order and does not have accessible text</>
<yellow></>
<yellow>Fix any of the following:</>
<yellow>  Element does not have text that is visible to screen readers</>
<yellow>  aria-label attribute does not exist or is empty</>
<yellow>  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty</>
<yellow>  Element has no title attribute</>

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/4.4/link-name?application=axeAPI</>

Expected the HTML found at $('a[href$=\\"#link-name-2\\"]') to have no violations:

<gray><a href=\\"#link-name-2\\"></a></>

Received:

<red>\\"Links must have discernible text (link-name)\\"</>

<yellow>Fix all of the following:</>
<yellow>  Element is in tab order and does not have accessible text</>
<yellow></>
<yellow>Fix any of the following:</>
<yellow>  Element does not have text that is visible to screen readers</>
<yellow>  aria-label attribute does not exist or is empty</>
<yellow>  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty</>
<yellow>  Element has no title attribute</>

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/4.4/link-name?application=axeAPI</>

────────

Expected the HTML found at $('img[src$=\\"example.com\\"]') to have no violations:

<gray><img src=\\"http://example.com\\"></>

Received:

<red>\\"All page content should be contained by landmarks (region)\\"</>

<yellow>Fix any of the following:</>
<yellow>  Some page content is not contained by landmarks</>

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/4.4/region?application=axeAPI</>

Expected the HTML found at $('img[src=\\"http://example.com/2\\"]') to have no violations:

<gray><img src=\\"http://example.com/2\\"></>

Received:

<red>\\"All page content should be contained by landmarks (region)\\"</>

<yellow>Fix any of the following:</>
<yellow>  Some page content is not contained by landmarks</>

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/4.4/region?application=axeAPI</>

Expected the HTML found at $('input') to have no violations:

<gray><input type=\\"text\\"></>

Received:

<red>\\"All page content should be contained by landmarks (region)\\"</>

<yellow>Fix any of the following:</>
<yellow>  Some page content is not contained by landmarks</>

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/4.4/region?application=axeAPI</>"
`;
